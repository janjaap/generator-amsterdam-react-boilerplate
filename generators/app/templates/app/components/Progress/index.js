import React, { useEffect, createRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { determined, undetermined, label as labelStyles } from './styles';

const Wrapper = styled(({ className, children }) => <div className={className}>{children}</div>)`
  ${({ type }) => (type === 'determined' ? determined.Wrapper : undetermined.Wrapper)};
`;

const Pie = styled(({ className }) => <div className={className} />)`
  ${({ type }) => (type === 'determined' ? determined.Pie : undetermined.Pie)};
`;

export const Label = styled.div`
  ${labelStyles}
`;

export const smallFactor = 0.333;

export const removeNode = ({ currentTarget }) => {
  if (currentTarget.classList.contains('finished')) {
    currentTarget.classList.add('hidden');
  }
};

/**
 * Fixed positioned component that indicates (loading) progress
 * Can be used to show page loading progress and will take a value for its `now` prop (between 0 and 1) to update the loader bar.
 */
const Progress = ({ className, label, labelPosition, showLabel, variant, ...props }) => {
  const scaleFactor = variant === 'small' ? smallFactor : 1;
  const progressRef = createRef();

  useEffect(() => {
    const { current } = progressRef;

    current.addEventListener('transitionend', removeNode, true);

    return () => {
      current.removeEventListener('transitionend', removeNode);
    };
  });

  return (
    <Wrapper
      ref={progressRef}
      className={`no-print ${className}${props.now >= 1 ? ' finished' : ''}`}
      scaleFactor={scaleFactor}
      as="div"
      data-testid="progress-wrapper"
      {...props}
    >
      <Pie key={Math.random()} scaleFactor={scaleFactor} {...props} />
      {showLabel && (
        <Label data-testid="progress-label" vPos={labelPosition}>
          {label !== undefined ? label : `${Math.round(props.now * 100)} %`}
        </Label>
      )}
    </Wrapper>
  );
};

Progress.defaultProps = {
  className: '',
  hasPrimaryColor: true,
  inline: false,
  label: undefined,
  labelPosition: 'middle',
  now: 0,
  showLabel: true,
  type: 'determined',
  variant: 'big',
};

Progress.propTypes = {
  /** @ignore */
  className: PropTypes.string,
  /** When false, the loader bar will inherit the color (through `currentColor`) from the parent (styled) component. */
  hasPrimaryColor: PropTypes.bool,
  /** When set to true, the Progress component won't be rendered as fixed positioned */
  inline: PropTypes.bool,
  /** Text label. When left empty, the relative progress (x %) is shown. */
  label: PropTypes.string,
  /** Vertical position of the text label */
  labelPosition: PropTypes.oneOf(['top', 'middle', 'bottom']),
  /** Relative progress. Should be a number between 0 and 1 */
  now: (props, propName) => {
    const value = props[propName];

    if (value < 0 || value > 1) {
      return new Error(`The value of 'now' should be between 0 and 1. Got ${value}.`);
    }

    return null;
  },
  /** Whether or not to show the (default) label */
  showLabel: PropTypes.bool,
  /** When type is set to 'undetermined', the component shows an infinite spinner. 'determined' Should be used in conjunction with `now`. */
  type: PropTypes.oneOf(['determined', 'undetermined']),
  /** Size indicator */
  variant: PropTypes.oneOf(['small', 'big']),
};

export default Progress;
