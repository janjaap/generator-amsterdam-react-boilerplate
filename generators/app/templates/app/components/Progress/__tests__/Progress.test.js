import React from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-styled-components';
import Progress, { Label, smallFactor, removeNode } from '..';
import { progressWidth } from '../styles';

describe('Progress', () => {
  afterEach(cleanup);

  it('matches the snapshot', () => {
    const { container } = render(<Progress />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('applies a scale factor for small variant', () => {
    const { rerender, getByTestId } = render(<Progress />);
    expect(getByTestId('progress-wrapper').firstChild).toHaveStyleRule('height', `${progressWidth}px`);
    expect(getByTestId('progress-wrapper').firstChild).toHaveStyleRule('width', `${progressWidth}px`);

    rerender(<Progress variant="small" />);
    expect(getByTestId('progress-wrapper').firstChild).toHaveStyleRule('height', `${progressWidth * smallFactor}px`);
    expect(getByTestId('progress-wrapper').firstChild).toHaveStyleRule('width', `${progressWidth * smallFactor}px`);
  });

  it('renders the progress element with the primary color', () => {
    const { rerender, getByTestId } = render(<Progress />);

    expect(getByTestId('progress-wrapper').firstChild).toHaveStyleRule('border-color', '#ec0000', {
      modifier: '::before',
    });

    rerender(<Progress hasPrimaryColor={false} />);

    expect(getByTestId('progress-wrapper').firstChild).toHaveStyleRule('border-color', 'currentColor', {
      modifier: '::before',
    });

    rerender(<Progress type="undetermined" />);

    expect(getByTestId('progress-wrapper').firstChild).toHaveStyleRule('color', '#ec0000');

    rerender(<Progress hasPrimaryColor={false} type="undetermined" />);

    expect(getByTestId('progress-wrapper').firstChild).toHaveStyleRule('color', 'currentColor');
  });

  it('renders the progress element as an inline element', () => {
    const { rerender, getByTestId } = render(<Progress />);

    expect(getByTestId('progress-wrapper')).toHaveStyleRule('position', 'fixed');

    rerender(<Progress inline />);

    expect(getByTestId('progress-wrapper')).toHaveStyleRule('position', 'relative');
  });

  it('renders a label', () => {
    const { rerender, getByText, getByTestId } = render(<Progress label="Foo bar" />);

    expect(getByText('Foo bar')).not.toBeUndefined();

    rerender(<Progress label="Foo bar" showLabel={false} />);

    expect(() => {
      getByText('Foo bar');
    }).toThrow();

    rerender(<Progress />);

    expect(getByTestId('progress-label').textContent).toContain(0);
  });

  it('shows the progress percentage', () => {
    const { rerender, getByTestId } = render(<Progress />);

    expect(getByTestId('progress-label').textContent).toBe('0 %');

    rerender(<Progress now={0.15} />);

    expect(getByTestId('progress-label').textContent).toBe('15 %');

    global.console.error = jest.fn();

    rerender(<Progress now={15} />);

    expect(global.console.error).toHaveBeenCalledWith(expect.stringContaining('should be between 0 and 1'));

    global.console.error.mockRestore();
  });

  describe('removeNode', () => {
    it('should add a classname', () => {
      const currentTarget = document.createElement('div');
      currentTarget.classList.add('some-class');

      const event = {
        currentTarget,
      };

      removeNode(event);

      expect(currentTarget.classList).not.toContain('hidden');

      currentTarget.classList.add('finished');

      removeNode(event);

      expect(currentTarget.classList).toContain('hidden');
    });
  });

  describe('label', () => {
    it('should position the label text vertically', () => {
      const { container, rerender } = render(<Label />);

      expect(container.firstChild).toHaveStyleRule('align-self', 'center');

      rerender(<Label vPos="center" />);

      expect(container.firstChild).toHaveStyleRule('align-self', 'center');

      rerender(<Label vPos="top" />);

      expect(container.firstChild).toHaveStyleRule('align-self', 'flex-start');

      rerender(<Label vPos="bottom" />);

      expect(container.firstChild).toHaveStyleRule('align-self', 'flex-end');
    });
  });
});
