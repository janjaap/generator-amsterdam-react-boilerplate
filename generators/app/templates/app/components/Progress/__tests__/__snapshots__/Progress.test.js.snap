// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress matches the snapshot 1`] = `
.c0 {
  background-color: #fff;
  border: 1px solid #000;
  border-top: 0;
  position: fixed;
  left: calc(50% - 50px);
  z-index: 500;
  box-sizing: content-box;
  -webkit-transition: -webkit-transform 0.35s ease-out;
  -webkit-transition: transform 0.35s ease-out;
  transition: transform 0.35s ease-out;
  -webkit-transform: translate3D(0,0px,0);
  -ms-transform: translate3D(0,0px,0);
  transform: translate3D(0,0px,0);
  top: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100px;
  will-change: transform;
}

.c0.finished {
  -webkit-transition-delay: 0.7s;
  transition-delay: 0.7s;
}

.c0.hidden {
  display: none;
}

.c1 {
  height: 60px;
  width: 60px;
  position: absolute;
  -webkit-clip: rect(0,60px,60px,30px);
  clip: rect(0,60px,60px,30px);
  will-change: clip;
}

.c1::before,
.c1::after {
  border-radius: 50%;
  border: 6px solid;
  border-color: #ec0000;
  -webkit-clip: rect(0,30px,60px,0);
  clip: rect(0,30px,60px,0);
  content: '';
  position: absolute;
  -webkit-transform: translate3D(0,0,0);
  -ms-transform: translate3D(0,0,0);
  transform: translate3D(0,0,0);
  -webkit-transition: -webkit-transform 0.1s ease-in;
  -webkit-transition: transform 0.1s ease-in;
  transition: transform 0.1s ease-in;
  height: 48px;
  width: 48px;
  will-change: transform;
}

.c1::before {
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c1::after {
  display: none;
}

.c2 {
  text-align: center;
  font-size: 14px;
  height: 18px;
  line-height: 18px;
  color: initial;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

@media screen and (-ms-high-contrast:active),(-ms-high-contrast:none) {
  .c1 {
    top: calc(50% - 30px);
    left: calc(50% - 30px);
  }
}

<div
  class="no-print  c0"
  data-testid="progress-wrapper"
  type="determined"
>
  <div
    class="c1"
  />
  <div
    class="c2"
    data-testid="progress-label"
  >
    0 %
  </div>
</div>
`;
